Add MAG IST8310

McGiverGim opened the pull request:
This adds support to the IST8310 compass.

Fixes https://github.com/betaflight/betaflight/issues/5386

Is based in the code of iNav https://github.com/iNavFlight/inav/blob/master/src/main/drivers/compass/compass_ist8310.c

ledvinap commented on src/main/drivers/compass/compass_ist8310.c:
```diff
@@ -0,0 +1,166 @@
+static bool ist8310Read(magDev_t *magDev, int16_t *magData)
+{
+    uint8_t buf[6];
+    uint8_t LSB2FSV = 3; // 3mG - 14 bit
```
(`const int`)

ledvinap commented on src/main/drivers/compass/compass_ist8310.c:
```diff
@@ -0,0 +1,166 @@
+    busDevice_t *busdev = &magDev->busdev;
+
+    // set magData to zero for case of failed read
+    magData[X] = 0;
```
Can be moved to `!ack` branch ...

ledvinap commented on src/main/drivers/compass/compass_ist8310.c:
```diff
@@ -0,0 +1,166 @@
+        busdev->busdev_u.i2c.address = IST8310_ADDRESS;
+    }
+
+    delay(10);
```
Is this delay necessary?

McGiverGim replied:
I don't think so, I will remove it and ask the user to test it again.

McGiverGim commented:
Corrections done as requested.

Only one thing:
- Tested in OMINBUSF4SD, all is ok, no I2C errors and the MAG seems to work.
- Tested in OMNIBUSF7V2, the I2C error count in the bottom maks 3 error, it does not increase, and the MAG seems to work.

I don't know why the F7 shows these errors, I don't have the compass to make tests by my own. Some idea?

McGiverGim commented:
Some more tests give some strange results in the X axis (it seems a problem of calibration). More info here https://github.com/betaflight/betaflight/issues/5386

I don't have the hardware, and it seems I can't find the problem. Someone with this compass can take a look to see where is the problem?

mikeller commented:
@McGiverGim: Are you still working on this, or is this ready to be merged?

McGiverGim commented:
@mikeller  I have some problems, but not only with the IST8310, the QMC5883 is doing strange things too. Maybe something is broken with the calibration or the mag code itself. I need more testing.

McGiverGim commented:
@mikeller if my final thoughts of https://github.com/betaflight/betaflight/pull/5451 are true, I think this is ready to be merged.

McGiverGim commented:
Again with doubts. Sorry for the yes/no game.

By the moment, don't merge it. We have doubts with some strange values.

mikeller commented:
All good, no hurry, we do not have any looming deadlines.

As a recommendation, can you use labels like 'Don't merge' or 'In development' to indicate the status of pull requests? This makes it easy to go through the list of pull requests and see what is ready and what not.

McGiverGim commented:
@mikeller I don't have permissions to add labels ;)

mikeller commented:
Now you do.

McGiverGim commented:
Isentek (the manufacturer) send me some changes to the init code to see if they solve the issue but don't, but I add them here to not lose it. It adds a soft reset and a pulse instruction in the init function:
```C
// I2C Control Register
#define IST8310_REG_PDCNTL  0x42

// sensor pulse duration setting
#define IST8310_PDCNTL_NORMAL   0xC0
#define IST8310_PDCNTL_LONG     0x40

static bool ist8310Init(magDev_t * magDev)
{
    busDevice_t *busdev = &magDev->busdev;
    uint8_t regTemp;

    // soft-Reset
    bool ack = busReadRegisterBuffer(busdev, IST8310_REG_CNTRL2, &regTemp, 1);
    regTemp |= IST8310_CNTRL2_RESET;
    ack = ack && busWriteRegister(busdev, IST8310_REG_CNTRL2, regTemp);
    delay(30);

    // ODR mode
    ack = ack && busWriteRegister(busdev, IST8310_REG_CNTRL1, IST8310_ODR_50_HZ);
    delay(5);

    // Init setting : avg16 / pulse mode
    ack = ack && busWriteRegister(busdev, IST8310_REG_AVERAGE, IST8310_AVG_16);
    ack = ack && busWriteRegister(busdev, IST8310_REG_PDCNTL, IST8310_PDCNTL_NORMAL);
    delay(5);

    return ack;
}
```

McGiverGim commented:
Only to put this PR up to date. After some tests, we have detected that the problem is that the second byte read in continuous mode, for some reason, gets corrupted. Some reads are right, others are corrupted. Seems some kind of timing issue or similar.

We have find two workarounds for this:
- The first is not to use the continuous mode. If we use the single mode (we write to the bus asking for a new value before each MAG read) it works and nothing gets corrupted.
- The second, is a very ugly hack. The second byte is corrupted, so in place of asking for 6 bytes, we can ask for 8 bytes, starting for the (X axis - 2) position and discard the two first bytes. The other 6 bytes contains the X, Y and Z positions without corruption.

I don't know if single mode is a valid solution. The second maybe is showing some little bug at some part of the I2C system but I don't have the tools/knowledge to detect it. In iNav it works right, without any hack.

McGiverGim commented:
Up to date about this:
- I have changed the MATEKF405 CTR by a MATEKF722
- I've rebased against master, there are some changes in the I2C bus and I wanted to see if some of them fixed the issue.

Result, now the bug has going worst:
- Under continuous mode, all three axis are now affected by the bug and not only the X axis.
- Under single mode, the X axis is now affected too by the bug. The Y and Z axis work right.

We have tested before in other F7 boards, so I think the culprit is possibly the new code added to master but is not sure.

I don't know if is better to close this PR until something happens that can fix it, or let it open to see if someone has an idea about this.

Ralfde commented:
How did inav solve this problem with the IST8310? with 2.0.0 RC4 all problems with IST8310 seem to be solved.

McGiverGim commented:
INav and Betaflight have different I2C code, different bus speed, etc. So I don't know exactly the difference that make it work. And I don't have the knowledge to trace it. The driver is the same.

Ralfde commented:
Is the different Bus speed essential for Betaflight? Should we make a test with changed Bus speed to see if it would solve the problem? If it would solve the problem, then maybe the BUS speed could be changed or made adjustable by CLI or whatever.

mikeller commented:
@Ralfde: Part of the design considerations in Betaflight, but not in iNav, is to enable users with boards with I2C attached gyros to have as fast as possible loop speeds, and that is what Betaflight is optimised for.

Ralfde commented:
Maybe use a different I2C handling for all boards, which use SPI for the gyro, which should be the majority in the meantime, especially as the F1 boards are not supported any more anyway. How many boards are left, which still use the I2C gyro?

mikeller commented:
@Ralfde: There is SPRACINGF3 and a couple of the other F3 boards that still support I2C gyros (even though they often support I2C *or* SPI). But with SPRACINGF3 still being the most downloaded Betaflight target, this isn't going to go away anytime soon...

Ralfde commented:
In this case maybe the software could check, if a I2C gyro is used or not and if not use the other I2C configuration for the compass (could be a compiler or software option)

McGiverGim commented:
My feature request (link in my other message) lets the user select it in CLI, what is the same. But as I say, this is NOT enough for this compass to work. Other things must be different between iNav and Betaflight. 
When changed the bus speed, the data gets corrupted for the first two bytes, as commented in other of my previous messages.
So the only solution, is that someone with more knowledge, time and the hardware take a look to this debugging the I2C bus to see what is happening.

McGiverGim commented:
I will close this PR. The problem remains and it can stay in that stay for too long... I will check from time to time with the new code, specially when there are changes in the I2C bus, to see if it fixes the problem. When fixed, I will open it again.
